File: cdm/cpp/io/protobuf/PBConditions.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "io/protobuf/PBConditions.h"
#include "io/protobuf/PBPatientConditions.h"
#include "io/protobuf/PBEnvironmentConditions.h"
#include "bind/cdm/Conditions.pb.h"
#include "bind/cdm/Scenario.pb.h"
#include "patient/conditions/SEPatientCondition.h"
#include "system/environment/conditions/SEEnvironmentCondition.h"
#include "substance/SESubstanceManager.h"


SECondition* PBCondition::Load(const cdm::AnyConditionData& condition, SESubstanceManager& subMgr)
{
  switch (condition.Condition_case())
  {
  case cdm::AnyConditionData::kPatientCondition:
    return PBPatientCondition::Load(condition.patientcondition(), subMgr);
  case cdm::AnyConditionData::kEnvironmentCondition:
    return PBEnvironmentCondition::Load(condition.environmentcondition(), subMgr);
  }
  subMgr.Error("Unknown Condition");
  return nullptr;
}
cdm::AnyConditionData* PBCondition::Unload(const SECondition& condition)
{
  cdm::AnyConditionData* any = new cdm::AnyConditionData();
  const SEPatientCondition* pc = dynamic_cast<const SEPatientCondition*>(&condition);
  if (pc != nullptr)
  {
    any->set_allocated_patientcondition(PBPatientCondition::Unload(*pc));
    return any;
  }
  const SEEnvironmentCondition* ec = dynamic_cast<const SEEnvironmentCondition*>(&condition);
  if (ec != nullptr)
  {
    any->set_allocated_environmentcondition(PBEnvironmentCondition::Unload(*ec));
    return any;
  }
  condition.Error("Unsupported Condition");
  delete any;
  return nullptr;
}

void PBCondition::Serialize(const cdm::ConditionData& src, SECondition& dst)
{
  dst.Clear();
  dst.SetComment(src.comment());
}
void PBCondition::Serialize(const SECondition& src, cdm::ConditionData& dst)
{
  dst.set_comment(src.m_Comment);
}

File: cdm/csharp/io/PBConditions.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

using System.Collections.Generic;
using Google.Protobuf;

namespace Pulse.CDM
{
  public class PBCondition
  {
    public static void Load(Cdm.ConditionListData src, List<SECondition> dst)
    {
      Serialize(src, dst);
    }
    public static void Serialize(Cdm.ConditionListData src, List<SECondition> dst)
    {
      foreach (var Condition in src.AnyCondition)
        dst.Add(PBCondition.Load(Condition));
    }
    public static bool SerializeFromString(string src, List<SECondition> dst)
    {
      try
      {
        Cdm.ConditionListData data = JsonParser.Default.Parse<Cdm.ConditionListData>(src);
        PBCondition.Load(data, dst);
      }
      catch (Google.Protobuf.InvalidJsonException)
      {
        return false;
      }
      return true;
    }

    public static Cdm.ConditionListData Unload(List<SECondition> src)
    {
      Cdm.ConditionListData dst = new Cdm.ConditionListData();
      Serialize(src, dst);
      return dst;
    }
    public static void Serialize(List<SECondition> src, Cdm.ConditionListData dst)
    {
      foreach (var Condition in src)
        dst.AnyCondition.Add(PBCondition.Unload(Condition));
    }
    public static string SerializeToString(List<SECondition> src)
    {
      var pb = PBCondition.Unload(src);
      return pb.ToString();
    }

    #region AnyCondition
    /** Create a new Condition based on the binding object,
     *  load that data into the new Condition, and return said Condition */
    public static SECondition Load(Cdm.AnyConditionData Condition)
    {
      switch (Condition.ConditionCase)
      {
        case Cdm.AnyConditionData.ConditionOneofCase.PatientCondition:
          return PBPatientCondition.Load(Condition.PatientCondition);
          //case Cdm.AnyConditionData.ConditionOneofCase.EnvironmentCondition:
          //  return PBEnvironmentCondition.Load(Condition.EnvironmentCondition);
      }
      return null;
    }
    /** Create a new bind object, unload the Condition,
     *  put that in the bind object, and return said bind object */
    public static Cdm.AnyConditionData Unload(SECondition Condition)
    {
      Cdm.AnyConditionData any = new Cdm.AnyConditionData();
      if (Condition.GetType().IsSubclassOf(typeof(SEPatientCondition)))
        any.PatientCondition = PBPatientCondition.Unload((SEPatientCondition)Condition);
      //if (Condition.GetType().IsSubclassOf(typeof(SEEnvironmentCondition)))
      //  any.EnvironmentCondition = PBEnvironmentCondition.Unload((SEEnvironmentCondition)Condition);
      return any;
    }
    #endregion

    #region SECondition
    public static void Serialize(Cdm.ConditionData src, SECondition dst)
    {
      dst.Clear();
      if (src.Comment != null)
        dst.SetComment(src.Comment);
    }
    public static void Serialize(SECondition src, Cdm.ConditionData dst)
    {
      if (src.HasComment())
        dst.Comment = src.GetComment();
    }
    #endregion
  }
}
